### PR TITLE
[autofill] opt-out instead of opt-in

### LINKED ISSUE #85554: Let Autofill be Opt-out
As per the request of the AwG team, we'd like to make autofill on Android to be opt-out to maintain parity with the existing Android Platform. In the long term, we might want to send all the information in the widget hierarchy to the autofill service. For now the following info of a text field should be sent to get a reasonable detection accuracy:

- [ ] placeholder text (https://github.com/flutter/flutter/pull/86312)
- [x] text
- [x] size & coordinates (should be implemented already)
- [ ] developer identifier (not sure how)
- [ ] input type (already available on android & iOS, should be trivial to add)
- [x] autofill type (current autofill is for text only)

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
- fixes https://github.com/flutter/flutter/issues/85554

To opt-out, specify a null `autofillHints`.

Additionally, allows high-level text input widgets to pass an
`AutofillClient` to `EditableText`. This allows these widgets to provide
additional autofill information that `EditableTextState` usually does
not have access to (for instance, the placeholder text).

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/86312

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 1]
```diff
@@ -1076,6 +1076,28 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with Restoratio
...
+
+  @override
+  TextInputConfiguration get textInputConfiguration {
+    final List<String>? autofillHints = widget.autofillHints?.toList(growable: false);
+    final AutofillConfiguration autofillConfiguration = autofillHints != null
+      ? AutofillConfiguration(
+          uniqueIdentifier: autofillId,
+          autofillHints: autofillHints,
+          currentEditingValue: _effectiveController.value,
+          hintText: widget.placeholder,
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:1094]: What is hintText actually used for, I guess somewhere in the engine already?  I understand autofillHints is the main list of hints that's used by the platform to determine what to autofill.

**Review by justinmc — APPROVED** [review]
LGTM.  One question about hintText.  I really think making this opt-out is a great idea, it should make many Flutter apps much more likely to have working autofill.

**Comment by justinmc** [comment]
@LongCatIsLooong I think the Linux failures may be infrastructure flakes?  Not sure about Google testing failures.  Also you have conflicts FYI.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/text_field.dart:1094]: AFAIK currently we don't send the hint text. The reason the text input plugin needs it is when the autofill hints list is empty it's a good indication for what the text field is actually for: https://github.com/flutter/flutter/issues/85554

**Pull request merged**